### PR TITLE
feat(tests): add call memory expansion with no return size test

### PR DESCRIPTION
**Pull request by JereSalo**
## 🗒️ Description

Test that when calling a contract with empty return size but an offset larger than the actual memory size the memory doesn't have to be expanded. One of the ways in which we can check if it has been expanded is for example by seeing the cost of an MSTORE, if it was expanded then it will cost less than what it should.

---
Hi guys! I just made a simple test for a bug that we had during a refactor in un our EVM implementation on ethrex and we caught it while syncing Hoodi and Holesky. E.g. [this holesky transaction](https://holesky.etherscan.io/tx/0xb3d2ac17c47a8c2fcb9378067951aad5edd84da726b076bea692b29382c29953) or [this hoodi transaction](https://hoodi.etherscan.io/tx/0x06594d4e429292c858003fa0b5ae3efd0ca664eb3a58781940d7d7b048377ae9)
I know it can be improved so I’d love advice for making this better. Specially the gas calculations aspect of it.
I had a few questions though:
- Do you accept test additions like this? I mean it is for a very particular edge case that happened within our EVM implementation, is it welcome?
- Where does this test belong? I left it in an incorrect file because I didn’t really know what the appropriate file for this test was.

Any suggestions are welcome, and if you feel like this test shouldn’t be added for some reason there’s no problem! Just wanted to know your opinion. Thanks!
I tested it with our EVM implementation and it detected the bug that we had, now after the fix it passes.

## 🔗 Related Issues or PRs

On ethrex: https://github.com/lambdaclass/ethrex/pull/3774

## ✅ Checklist

- [ ] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx --with=tox-uv tox -e lint,typecheck,spellcheck,markdownlint
    ```
- [ ] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [ ] All: Considered adding an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [ ] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-spec-tests/blob/main/docs/writing_tests/post_mortems.md) to add an entry the list.
- [ ] Ported Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) or [tests/static](/ethereum/execution-spec-tests/blob/main/tests/static) have been assigned `@ported_from` marker.

### CODE REVIEW

File: tests/prague/eip7702_set_code_tx/test_set_code_txs_2.py [hunk 1]
```diff
@@ -1806,3 +1806,54 @@ def test_delegation_replacement_call_previous_contract(
...
+    gsc = fork.gas_costs()
+    overhead_cost = gsc.G_VERY_LOW * (len(Op.CALL.kwargs) + len(Op.MSTORE.kwargs))  # type: ignore
+    mem_offset = 128  # arbitrary number
+
+    contract_code = CodeGasMeasure(
+        code=Op.CALL(gas=0, ret_offset=mem_offset, ret_size=0)
+        + Op.MSTORE(offset=mem_offset, value=1),
+        overhead_cost=overhead_cost,
+        extra_stack_items=1,  # Because CALL pushes 1 item to the stack
+    )
```
Comment (marioevz) [inline tests/prague/eip7702_set_code_tx/test_set_code_txs_2.py:1833]: I think that `Op.CALL` and `Op.MSTORE` would have to go each in a separate `CodeGasMeasure`, each writing to a separate storage, to catch the difference in gas for the memory expansion only happening in the `Op.MSTORE` result.

File: tests/prague/eip7702_set_code_tx/test_set_code_txs_2.py
[the remark this answers is not in the document]
Comment (JereSalo) [reply tests/prague/eip7702_set_code_tx/test_set_code_txs_2.py:1]: Done [here](https://github.com/ethereum/execution-spec-tests/pull/1952/commits/808bf26fddb11553c2a6f4bc15691b35f6e18a27)

File: tests/prague/eip7702_set_code_tx/test_set_code_txs_2.py [hunk 2]
```diff
@@ -1806,3 +1806,54 @@ def test_delegation_replacement_call_previous_contract(
...
+        gas_limit=500_000,
+        to=contract,
+        value=0,
+        sender=sender,
+    )
+
+    # this call cost is just the address_access_cost
+    call_cost = gsc.G_COLD_ACCOUNT_ACCESS
+    # mstore cost: base cost + expansion cost
+    mstore_cost = gsc.G_MEMORY + (gsc.G_MEMORY * (mem_offset // 32 + 1))
```
[the remark this answers is not in the document]
Comment (JereSalo) [reply tests/prague/eip7702_set_code_tx/test_set_code_txs_2.py]: Nice! Thanks for the tip.
https://github.com/ethereum/execution-spec-tests/pull/1952/commits/78e4051a1db226f9ba8728983c4bb30f02e88839

Now I'm having a problem with the cost of CALL, because in old forks the `address_access_cost` doesn't work. What is your recommendation for this?

[the remark this answers is not in the document]
Comment (marioevz) [reply tests/prague/eip7702_set_code_tx/test_set_code_txs_2.py]: I'm taking a look at our framework and this seems like an issue with the lack of gas definitions for forks prior to Berlin.
In order to not increase the scope of this PR, I would simply constrain the test to only be filled for forks Berlin and after:
```python
@pytest.mark.valid_from("Berlin")
def test_call_large_offset_mstore(
```

**Comment by JereSalo** [comment]
> The transaction issue can be fixed using the `protected` field to be `True` when filling for Byzantium or after:
> 
> ```python
> tx = Transaction(
>         gas_limit=500_000,
>         to=contract,
>         value=0,
>         sender=sender,
>         protected=True if fork >= Byzantium else False,
>     )
> ```
> 
> But I think this is unnecessary if we do the workaround in my previous comment. Let's just mark the test to be filled only on Berlin and after for now.

Great! Added the Berlin check with a comment. Let me know if I should change anything else! Thanks
